File: problems/two_sum.py
'''
Given an array of integers nums and an integer target, return the indices i and j such that nums[i] + nums[j] == target and i != j.

You may assume that every input has exactly one pair of indices i and j that satisfy the condition.

Return the answer with the smaller index first.

Example 1:

Input: 
nums = [3,4,5,6], target = 7

Output: [0,1]
Explanation: nums[0] + nums[1] == 7, so we return [0, 1].

Example 2:

Input: nums = [4,5,6], target = 10

Output: [0,2]
Example 3:

Input: nums = [5,5], target = 10

Output: [0,1]
Constraints:

2 <= nums.length <= 1000
-10,000,000 <= nums[i] <= 10,000,000
-10,000,000 <= target <= 10,000,000
'''

# I have already solved this problem using brute force method. However, it was o(n**2).

# Approach 2: Using dictionary to keep track of all the elements in the given array

def twoSum(nums: list[int], target: int):
    nums_dict: dict[int, int] = {}
    
    for i in range(len(nums)):
        nums_dict[nums[i]] = i
    
    for i in range(len(nums)):
        diff = target - nums[i]
        
        if nums_dict.get(diff):
            found = nums_dict[diff]
            
            if found == i:
                continue
            
            return [i, found] if i < found else [found, i]

    return False

if __name__ == "__main__":
    
    nums = [1,3,4,2,]
    target = 6
    
    print(twoSum(nums, target))

File: arrays/two_sum.cpp
/*
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 

Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
Example 2:

Input: nums = [3,2,4], target = 6
Output: [1,2]
Example 3:

Input: nums = [3,3], target = 6
Output: [0,1]
 

Constraints:

2 <= nums.length <= 104
-109 <= nums[i] <= 109
-109 <= target <= 109
Only one valid answer exists.
 

Follow-up: Can you come up with an algorithm that is less than O(n2) time complexity?

*/

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

// Attempt 1: Using a hashmap maybe?

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> umap;
        
        for (size_t i = 0; i < nums.size(); i++ ) {
            const int complement = target - nums[i];
            auto found = umap.find(complement);

            if (found != umap.end()) {
                return {found->second, (int) i};

            } else {
                umap[nums[i]] = i;
            }
        }

        return {-1, -1};
    }
};

int main() {
    Solution slu;
    vector<int> nums = {3, 2, 4};
    int target = 6;
    vector<int> answer = slu.twoSum(nums, target);

    cout << "Solution: " << endl;
    for (int i : answer) {
        cout << i << endl;
    }
    
    return 0;
}